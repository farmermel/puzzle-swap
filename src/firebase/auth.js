import { auth } from './firebase';

export const doCreateUserWithEmailAndPassword = (email, password) =>
  auth.createUserWithEmailAndPassword(email, password);

export const doSignInWithEmailAndPassword = (email, password) => 
  auth.signInWithEmailAndPassword(email, password);

export const onAuthStateChanged = user => 
  auth.onAuthStateChanged(user);

export const doSignOut = () => 
  auth.signOut();