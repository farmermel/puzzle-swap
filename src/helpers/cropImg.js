/* eslint-disable */
export const getCroppedImg = (image, pixelCrop, fileName) => {

  const canvas = document.createElement('canvas');
  canvas.width = pixelCrop.width;
  canvas.height = pixelCrop.height;
  const ctx = canvas.getContext('2d');
  console.log('image', image)
  console.log('pixelCropxy', pixelCrop.x, pixelCrop.y)
  ctx.drawImage(
    image,
    pixelCrop.x,
    pixelCrop.y,
    pixelCrop.width,
    pixelCrop.height,
    0,
    0,
    pixelCrop.width,
    pixelCrop.height
  );

  return new Promise((resolve, reject) => {
    canvas.toBlob(file => {
      file.name = fileName;
      resolve(file);
    }, 'image/jpeg');
  });
}