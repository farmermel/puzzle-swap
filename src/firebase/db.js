import { db } from './firebase';

export const doCreateUser = (id, username, email) => 
  db.ref(`users/${id}`).set({
    username,
    email
  });

// export const onceGetUsers = () => 
//   db.ref('users').once('value');

export const getOnce = path =>
  db.ref(path).once('value');

export const getFirebaseKey = childNode => 
  db.ref().child(childNode).push().key;

export const postUpdate = updates => 
  db.ref().update(updates);

export const watchData = watchNode => 
  db.ref(watchNode);

// export const getData = watchNode => 
//   watchData(watchNode).on('value', (snapshot) => {
//     return snapshot.val()
//   })