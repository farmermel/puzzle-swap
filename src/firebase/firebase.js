import * as firebase from 'firebase';
import { firebaseKey,
  fbAuthDomain,
  fbdbUrl,
  fbProjId,
  fbMessagingSenderId } from '../helpers/apiKeys';

const config = {
  apiKey: firebaseKey,
  authDomain: fbAuthDomain,
  databaseURL: fbdbUrl,
  projectId: fbProjId,
  storageBucket: "",
  messagingSenderId: fbMessagingSenderId
};

if (!firebase.apps.length) {
  firebase.initializeApp(config);
}

const auth = firebase.auth();

const db = firebase.database();

export {
  db,
  auth,
};