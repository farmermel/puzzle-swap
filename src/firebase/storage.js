import { storage } from './firebase';

export const getStoreRef = path =>
  storage.ref().child(path);

export const putInStore = (ref, blob) =>
  ref.put(blob);