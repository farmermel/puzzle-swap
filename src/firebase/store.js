import { store } from './firebase';

export const getStoreRef = path =>
  store.ref().child(path)

export const putInStore = (ref, blob) =>
  ref.put(blob)