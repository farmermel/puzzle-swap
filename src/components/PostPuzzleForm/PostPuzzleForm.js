import React, { Component } from 'react';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import { db, storage } from '../../firebase';
import ReactCrop from 'react-image-crop';
import { getCroppedImg } from '../../helpers/cropImg';
import 'react-image-crop/dist/ReactCrop.css';

export class PostPuzzleForm extends Component {
  constructor() {
    super();
    this.state = {
      title: '',
      numPieces: '',
      piecesMissing: '0',
      fileUpload: 'Select puzzle photo',
      crop: {
        aspect: 10/7,
        height: 35,
        width: 50,
        x: 0,
        y: 0
      } 
    }
  }

  handleChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  handlePhoto = (files: FileList) => {
    const puzzleImg = files[0];
    const imageUrl = URL.createObjectURL(puzzleImg);
    files[0] && this.setState({ 
                  imageUrl,
                  fileUpload: files[0].name
                })
  }

  postToDB = puzzleId => {
    const { userId, hasErrored, userLocation } = this.props;
    try {
      const { title, numPieces, piecesMissing } = this.state;
      const postDB = { title, numPieces, piecesMissing, puzzleId, userId, userLocation }
      const firebaseKey = db.getFirebaseKey('puzzles');
      let updates = {};
      updates[`/puzzles/${firebaseKey}`] = postDB;
      return db.postUpdate(updates);
    } catch (error) {
      hasErrored(error.message);
    }
  }

  postToCloudStore = () => {
    const { hasErrored } = this.props;
    try {
      const { croppedImg } = this.state;
      const puzzleId = Date.now();
      const ref = storage.getStoreRef(`images/${puzzleId}`);
      storage.putInStore(ref, croppedImg);
      return puzzleId;
    } catch (error) {
      hasErrored(error.message);
    }
  }

  postPuzzleToFirebase = async e => {
    e.preventDefault();
    try {
      const puzzleId = await this.postToCloudStore();
      await this.postToDB(puzzleId);
      this.props.history.push('/');
    } catch (error) {
      const { hasErrored } = this.props;
      hasErrored(error.message);
    }
  }

  handleCropChange = crop => {
    this.setState({ crop });
  }

  handleCropComplete = async (crop, pixelCrop) => {
    const { hasErrored } = this.props;
    try {
      const image = document.createElement('img');
      image.src = this.state.imageUrl;
      const croppedImgBlob = await getCroppedImg(image, pixelCrop, this.state.fileUpload);
      const croppedImg = new File([croppedImgBlob], this.state.fileUpload);
      this.setState({ croppedImg });
    } catch (error) {
      hasErrored(error.message);
    }
  }

  renderCrop = () => {
    return (
      <div className='crop-wrap'>
        <label htmlFor='crop-img'
               className='crop-label'>Crop your image</label>
        <ReactCrop src={ this.state.imageUrl }
                   crop={ this.state.crop }
                   onChange={ crop => this.handleCropChange(crop) }
                   onComplete={ (crop, pixelCrop) => this.handleCropComplete(crop, pixelCrop) }
                   id='crop-img' />
      </div>
    )
  }

  render() {
    return (
      <div className='form-wrapper'>
        {
          this.state.error && <p>{ this.state.error }</p>
        }
        <form onSubmit={ this.postPuzzleToFirebase }>
          <label htmlFor='puzzle-name'>Puzzle Name</label>
            <input type='text' placeholder='puzzle title'
                   value={ this.state.title }
                   name='title'
                   id='puzzle-name'
                   onChange={ this.handleChange }
                   required />
          <label htmlFor='puzzle-img'
                 id='img-submit-label'>{ this.state.fileUpload }</label>
            <input type='file' 
                   accept='image/x-png,image/jpeg'
                   name='puzzleImg' 
                   onChange={(e) => this.handlePhoto(e.target.files)}
                   id='puzzle-img'
                   required />
          {
            this.state.imageUrl && this.renderCrop()
          }
          <label htmlFor='num-pieces'>Number of pieces</label>
            <input type='number' placeholder='number of pieces'
                   value={ this.state.numPieces }
                   name='numPieces'
                   id='num-pieces'
                   onChange={ this.handleChange }
                   required />
          <label htmlFor='missing'>Number of pieces missing</label>
            <select name='piecesMissing'
                    id='missing'
                    onChange={this.handleChange}
                    required>
              <option value='0'>0</option>
              <option value='1-3'>1-3</option>
              <option value='4-9'>4-9</option>
              <option value='10 or more'>10 or more</option>
            </select>
          <button>Post puzzle</button>
        </form>
      </div>
    )
  }
}

PostPuzzleForm.propTypes = {
  userId: PropTypes.string.isRequired,
  hasErrored: PropTypes.func.isRequired,
  userLocation: PropTypes.string
}

export default withRouter(PostPuzzleForm);