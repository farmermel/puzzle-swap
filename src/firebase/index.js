import * as auth from './auth';
import * as db from './db';
import * as firebase from './firebase';
import * as store from './store';

export {
  auth,
  db,
  store,
  firebase,
};