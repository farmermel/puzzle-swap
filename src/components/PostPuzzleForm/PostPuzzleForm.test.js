import React from 'react';
import { shallow } from 'enzyme';
import { db, storage } from '../../firebase';
import { PostPuzzleForm } from './PostPuzzleForm';
import * as cropHelper from '../../helpers/cropImg';

describe('PostPuzzleForm', () => {
  let wrapper;
  const mockHistory = {
    push: jest.fn()
  }

  beforeEach(() => {
    wrapper = shallow(<PostPuzzleForm history={mockHistory}
                                      userId='4'
                                      hasErrored={jest.fn()} />);
  })

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('has default state', () => {
    const expected = {
      fileUpload: "Select puzzle photo", 
      numPieces: "", 
      piecesMissing: "0", 
      title: "",
      crop: {
        aspect: 10/7,
        height: 35,
        width: 50,
        x: 0,
        y: 0
      }
    }
    expect(wrapper.instance().state).toEqual(expected);
  })

  describe('handleChange', () => {
    it('sets state with event passed in', () => {
      const mockEvent = {
        target: {
          name: 'name',
          value: 'Nymeria'
        }
      }
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.instance().state.name).toEqual('Nymeria');
    })
  })

  describe('handlePhoto', () => {
    it('sets state with a File', () => {
      URL.createObjectURL = jest.fn().mockImplementation(() => {
        return 'I\'m an image'
      })
      const FileList = [
        {name: 'dragon puzzle'},
        {name: 'unicorn puzzle'}
      ]
      wrapper.instance().handlePhoto(FileList);
      expect(wrapper.instance().state.imageUrl).toEqual('I\'m an image');
      expect(wrapper.instance().state.fileUpload).toEqual('dragon puzzle');
    })
  })

  describe('postToDB', () => {
    beforeEach(() => {
      db.getFirebaseKey = jest.fn().mockImplementation(() => {
        return '5'
      })
      db.postUpdate = jest.fn()
    })

    it('calls getFirebaseKey method of database with pathway \'puzzles\' as argument', () => {
      expect(db.getFirebaseKey).not.toHaveBeenCalled();
      wrapper.instance().postToDB();
      expect(db.getFirebaseKey).toHaveBeenCalledWith('puzzles');
    })

    it('calls postUpdate method of database with an updates object as argument', () => {
      const updatesObj = {
        "/puzzles/5": 
        {
          numPieces: "", 
          piecesMissing: "0", 
          puzzleId: undefined, 
          title: "",
          userId: '4'
        }
      }
      expect(db.postUpdate).not.toHaveBeenCalled();
      wrapper.instance().postToDB();
      expect(db.postUpdate).toHaveBeenCalledWith(updatesObj);
    })

    it('catches error and calls hasErrored error message if anything errors', async () => {
      db.getFirebaseKey = jest.fn().mockImplementation(() => {
        throw new Error('failed to post to database')
      }); 
      await wrapper.instance().postToDB('4');
      expect(wrapper.instance().props.hasErrored).toHaveBeenCalledWith('failed to post to database');
    })
  })

  describe('postToCloudStore', () => {
    storage.getStoreRef = jest.fn().mockImplementation(() => {
      return 'ref'
    });
    storage.putInStore = jest.fn();
    Date.now = jest.fn().mockImplementation(() => {
      return 5
    })

    it('calls getStoreRef on storage with a url as parameter and putInStore with ref and puzzleimg as arguments', () => {
      wrapper.instance().state.croppedImg = 'puzzleImg';
      expect(storage.getStoreRef).not.toHaveBeenCalled();
      expect(storage.putInStore).not.toHaveBeenCalled();
      wrapper.instance().postToCloudStore();
      expect(storage.getStoreRef).toHaveBeenCalledWith('images/5');
      expect(storage.putInStore).toHaveBeenCalledWith('ref', 'puzzleImg');
    })

    it('returns a generated puzzleId', () => {
      expect(wrapper.instance().postToCloudStore()).toEqual(5)
    })

    it('catches error and calls hasErrored error message if anything errors', async () => {
      storage.getStoreRef = jest.fn().mockImplementation(() => {
        throw new Error('failed to post to database')
      }); 
      await wrapper.instance().postToCloudStore();
      expect(wrapper.instance().props.hasErrored).toHaveBeenCalledWith('failed to post to database');
    })
  })

  describe('postPuzzleToFirebase', () => {
    const mockEvent = {
      preventDefault: jest.fn()
    } 
    beforeEach(() => {
      wrapper.instance().postToCloudStore = jest.fn();
      wrapper.instance().postToDB = jest.fn(); 
    }) 

    it('calls prevent default on event passed in', () => {
      expect(mockEvent.preventDefault).not.toHaveBeenCalled();
      wrapper.instance().postPuzzleToFirebase(mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalled();
    })

    it('calls postToCloudStore', () => {
      expect(wrapper.instance().postToCloudStore).not.toHaveBeenCalled();
      wrapper.instance().postPuzzleToFirebase(mockEvent);
      expect(wrapper.instance().postToCloudStore).toHaveBeenCalled();
    })

    it('calls postToDB', async () => {
      expect(wrapper.instance().postToDB).not.toHaveBeenCalled();
      await wrapper.instance().postPuzzleToFirebase(mockEvent);
      expect(wrapper.instance().postToDB).toHaveBeenCalled();
    })

    it('calls push on history object with \'/\' as argument', async () => {
      wrapper.instance().props.history.push.mockClear();
      expect(wrapper.instance().props.history.push).not.toHaveBeenCalled();
      await wrapper.instance().postPuzzleToFirebase(mockEvent);
      expect(wrapper.instance().props.history.push).toHaveBeenCalledWith('/');
    })

    it('catches error and calls hasErrored error message if anything errors', async () => {
      wrapper.instance().postToDB = jest.fn().mockImplementation(() => {
        throw new Error('failed to post to database')
      }); 
      await wrapper.instance().postPuzzleToFirebase(mockEvent);
      expect(wrapper.instance().props.hasErrored).toHaveBeenCalledWith('failed to post to database');
    })
  })

  describe('handleCropChange', () => {
    it('sets state with crop passed in', () => {
      const mockStartCrop = {
        aspect: 10/7,
        height: 35,
        width: 50,
        x: 0,
        y: 0
      }
      const mockChangedCrop = {
        aspect: 10/7,
        height: 14,
        width: 20,
        x: 5,
        y: 10
      }
      expect(wrapper.instance().state.crop).toEqual(mockStartCrop);
      wrapper.instance().handleCropChange(mockChangedCrop);
      expect(wrapper.instance().state.crop).toEqual(mockChangedCrop);
    })
  })

  describe('handleCropComplete', () => {
    beforeAll(() => {
      cropHelper.getCroppedImg = jest.fn().mockImplementation(() => {
        return 'blob:http://localhose:3000/e4e674f8-4b25-4947-a46e-6b338b47d2f5'
      })
    })

    it('calls getCroppedImg with an html image element, a pixelCrop, and a filename', () => {
      expect(cropHelper.getCroppedImg).not.toHaveBeenCalled();
      wrapper.instance().handleCropComplete(wrapper.instance().state.crop, {width: 800, height: 900, x: 14, y: 4});
      expect(cropHelper.getCroppedImg).toHaveBeenCalled();
    })

    it('sets state with a croppedImg File', async () => {
      await wrapper.instance().handleCropComplete(wrapper.instance().state.crop, {width: 800, height: 900});
      expect(typeof wrapper.instance().state.croppedImg).toEqual('object');
    })

    it('catches error and calls hasErrored error message if anything errors', async () => {
      cropHelper.getCroppedImg = jest.fn().mockImplementation(() => {
        throw new Error('failed to post to database')
      }); 
      await wrapper.instance().handleCropComplete({}, {});
      expect(wrapper.instance().props.hasErrored).toHaveBeenCalledWith('failed to post to database');
    })
  })
})