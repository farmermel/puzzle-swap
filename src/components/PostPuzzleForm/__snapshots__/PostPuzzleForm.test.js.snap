// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostPuzzleForm matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostPuzzleForm
    history={
        Object {
            "push": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        <form
          onSubmit={[Function]}
>
          <label
                    htmlFor="puzzle-name"
          >
                    Puzzle Name
          </label>
          <input
                    id="puzzle-name"
                    name="title"
                    onChange={[Function]}
                    placeholder="puzzle title"
                    required={true}
                    type="text"
                    value=""
          />
          <label
                    htmlFor="puzzle-img"
                    id="img-submit-label"
          >
                    Select puzzle photo
          </label>
          <input
                    accept="image/x-png,image/jpeg"
                    id="puzzle-img"
                    name="puzzleImg"
                    onChange={[Function]}
                    required={true}
                    type="file"
          />
          <label
                    htmlFor="num-pieces"
          >
                    Number of pieces
          </label>
          <input
                    id="num-pieces"
                    name="numPieces"
                    onChange={[Function]}
                    placeholder="number of pieces"
                    required={true}
                    type="number"
                    value=""
          />
          <label
                    htmlFor="missing"
          >
                    Number of pieces missing
          </label>
          <select
                    id="missing"
                    name="piecesMissing"
                    onChange={[Function]}
                    required={true}
          >
                    <option
                              value="1-3"
                    >
                              1-3
                    </option>
                    <option
                              value="4-9"
                    >
                              4-9
                    </option>
                    <option
                              value="10 or more"
                    >
                              10 or more
                    </option>
          </select>
          <button>
                    Post puzzle
          </button>
</form>,
      ],
      "className": "form-wrapper",
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              htmlFor="puzzle-name"
>
              Puzzle Name
</label>,
            <input
              id="puzzle-name"
              name="title"
              onChange={[Function]}
              placeholder="puzzle title"
              required={true}
              type="text"
              value=""
/>,
            <label
              htmlFor="puzzle-img"
              id="img-submit-label"
>
              Select puzzle photo
</label>,
            <input
              accept="image/x-png,image/jpeg"
              id="puzzle-img"
              name="puzzleImg"
              onChange={[Function]}
              required={true}
              type="file"
/>,
            <label
              htmlFor="num-pieces"
>
              Number of pieces
</label>,
            <input
              id="num-pieces"
              name="numPieces"
              onChange={[Function]}
              placeholder="number of pieces"
              required={true}
              type="number"
              value=""
/>,
            <label
              htmlFor="missing"
>
              Number of pieces missing
</label>,
            <select
              id="missing"
              name="piecesMissing"
              onChange={[Function]}
              required={true}
>
              <option
                            value="1-3"
              >
                            1-3
              </option>
              <option
                            value="4-9"
              >
                            4-9
              </option>
              <option
                            value="10 or more"
              >
                            10 or more
              </option>
</select>,
            <button>
              Post puzzle
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Puzzle Name",
              "htmlFor": "puzzle-name",
            },
            "ref": null,
            "rendered": "Puzzle Name",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "puzzle-name",
              "name": "title",
              "onChange": [Function],
              "placeholder": "puzzle title",
              "required": true,
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select puzzle photo",
              "htmlFor": "puzzle-img",
              "id": "img-submit-label",
            },
            "ref": null,
            "rendered": "Select puzzle photo",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "accept": "image/x-png,image/jpeg",
              "id": "puzzle-img",
              "name": "puzzleImg",
              "onChange": [Function],
              "required": true,
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Number of pieces",
              "htmlFor": "num-pieces",
            },
            "ref": null,
            "rendered": "Number of pieces",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "num-pieces",
              "name": "numPieces",
              "onChange": [Function],
              "placeholder": "number of pieces",
              "required": true,
              "type": "number",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Number of pieces missing",
              "htmlFor": "missing",
            },
            "ref": null,
            "rendered": "Number of pieces missing",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="1-3"
>
                  1-3
</option>,
                <option
                  value="4-9"
>
                  4-9
</option>,
                <option
                  value="10 or more"
>
                  10 or more
</option>,
              ],
              "id": "missing",
              "name": "piecesMissing",
              "onChange": [Function],
              "required": true,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "1-3",
                  "value": "1-3",
                },
                "ref": null,
                "rendered": "1-3",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "4-9",
                  "value": "4-9",
                },
                "ref": null,
                "rendered": "4-9",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "10 or more",
                  "value": "10 or more",
                },
                "ref": null,
                "rendered": "10 or more",
                "type": "option",
              },
            ],
            "type": "select",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Post puzzle",
            },
            "ref": null,
            "rendered": "Post puzzle",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <form
            onSubmit={[Function]}
>
            <label
                        htmlFor="puzzle-name"
            >
                        Puzzle Name
            </label>
            <input
                        id="puzzle-name"
                        name="title"
                        onChange={[Function]}
                        placeholder="puzzle title"
                        required={true}
                        type="text"
                        value=""
            />
            <label
                        htmlFor="puzzle-img"
                        id="img-submit-label"
            >
                        Select puzzle photo
            </label>
            <input
                        accept="image/x-png,image/jpeg"
                        id="puzzle-img"
                        name="puzzleImg"
                        onChange={[Function]}
                        required={true}
                        type="file"
            />
            <label
                        htmlFor="num-pieces"
            >
                        Number of pieces
            </label>
            <input
                        id="num-pieces"
                        name="numPieces"
                        onChange={[Function]}
                        placeholder="number of pieces"
                        required={true}
                        type="number"
                        value=""
            />
            <label
                        htmlFor="missing"
            >
                        Number of pieces missing
            </label>
            <select
                        id="missing"
                        name="piecesMissing"
                        onChange={[Function]}
                        required={true}
            >
                        <option
                                    value="1-3"
                        >
                                    1-3
                        </option>
                        <option
                                    value="4-9"
                        >
                                    4-9
                        </option>
                        <option
                                    value="10 or more"
                        >
                                    10 or more
                        </option>
            </select>
            <button>
                        Post puzzle
            </button>
</form>,
        ],
        "className": "form-wrapper",
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                htmlFor="puzzle-name"
>
                Puzzle Name
</label>,
              <input
                id="puzzle-name"
                name="title"
                onChange={[Function]}
                placeholder="puzzle title"
                required={true}
                type="text"
                value=""
/>,
              <label
                htmlFor="puzzle-img"
                id="img-submit-label"
>
                Select puzzle photo
</label>,
              <input
                accept="image/x-png,image/jpeg"
                id="puzzle-img"
                name="puzzleImg"
                onChange={[Function]}
                required={true}
                type="file"
/>,
              <label
                htmlFor="num-pieces"
>
                Number of pieces
</label>,
              <input
                id="num-pieces"
                name="numPieces"
                onChange={[Function]}
                placeholder="number of pieces"
                required={true}
                type="number"
                value=""
/>,
              <label
                htmlFor="missing"
>
                Number of pieces missing
</label>,
              <select
                id="missing"
                name="piecesMissing"
                onChange={[Function]}
                required={true}
>
                <option
                                value="1-3"
                >
                                1-3
                </option>
                <option
                                value="4-9"
                >
                                4-9
                </option>
                <option
                                value="10 or more"
                >
                                10 or more
                </option>
</select>,
              <button>
                Post puzzle
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Puzzle Name",
                "htmlFor": "puzzle-name",
              },
              "ref": null,
              "rendered": "Puzzle Name",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "puzzle-name",
                "name": "title",
                "onChange": [Function],
                "placeholder": "puzzle title",
                "required": true,
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select puzzle photo",
                "htmlFor": "puzzle-img",
                "id": "img-submit-label",
              },
              "ref": null,
              "rendered": "Select puzzle photo",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "accept": "image/x-png,image/jpeg",
                "id": "puzzle-img",
                "name": "puzzleImg",
                "onChange": [Function],
                "required": true,
                "type": "file",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Number of pieces",
                "htmlFor": "num-pieces",
              },
              "ref": null,
              "rendered": "Number of pieces",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "num-pieces",
                "name": "numPieces",
                "onChange": [Function],
                "placeholder": "number of pieces",
                "required": true,
                "type": "number",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Number of pieces missing",
                "htmlFor": "missing",
              },
              "ref": null,
              "rendered": "Number of pieces missing",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    value="1-3"
>
                    1-3
</option>,
                  <option
                    value="4-9"
>
                    4-9
</option>,
                  <option
                    value="10 or more"
>
                    10 or more
</option>,
                ],
                "id": "missing",
                "name": "piecesMissing",
                "onChange": [Function],
                "required": true,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "1-3",
                    "value": "1-3",
                  },
                  "ref": null,
                  "rendered": "1-3",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "4-9",
                    "value": "4-9",
                  },
                  "ref": null,
                  "rendered": "4-9",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "10 or more",
                    "value": "10 or more",
                  },
                  "ref": null,
                  "rendered": "10 or more",
                  "type": "option",
                },
              ],
              "type": "select",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Post puzzle",
              },
              "ref": null,
              "rendered": "Post puzzle",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
